safekeeper/benches/receive_wal.rs

```diff
@@ -0,0 +1,153 @@
+//! WAL ingestion benchmarks.
+
+#[path = "benchutils.rs"]
+mod benchutils;
+
+use benchutils::Env;
+use criterion::{criterion_group, criterion_main, BatchSize, Bencher, Criterion};
+use itertools::Itertools as _;
+use postgres_ffi::v17::wal_generator::WalGenerator;
+use safekeeper::receive_wal::WalAcceptor;
+use safekeeper::safekeeper::{
+    AcceptorProposerMessage, AppendRequest, AppendRequestHeader, ProposerAcceptorMessage,
+};
+use utils::id::{NodeId, TenantTimelineId};
+use utils::lsn::Lsn;
+
+// Register benchmarks with Criterion.
+criterion_group!(benches, bench_process_msg, bench_wal_acceptor);
+criterion_main!(benches);
+
+/// Benchmarks SafeKeeper::process_msg() as time per message. Each message is an AppendRequest with
+/// a single WAL record containing a tiny XlLogicalMessage (~64 bytes total).
+fn bench_process_msg(c: &mut Criterion) {
+    let mut g = c.benchmark_group("process_msg");
+    for fsync in [false, true] {
+        g.bench_function(format!("fsync={fsync}"), |b| run_bench(b, fsync).unwrap());
+    }
+
+    // The actual benchmark.
+    fn run_bench(b: &mut Bencher, fsync: bool) -> anyhow::Result<()> {
+        let runtime = tokio::runtime::Builder::new_current_thread() // single is fine, sync IO only
+            .enable_all()
+            .build()?;
+
+        let env = Env::new(fsync)?;
+        let mut safekeeper =
+            runtime.block_on(env.make_safekeeper(NodeId(1), TenantTimelineId::generate()))?;
+        let mut walgen = WalGenerator::new();
+
+        b.iter_batched_ref(
+            // Pre-construct WAL records and requests. Criterion will batch them.
+            || {
+                let (lsn, record) = walgen.next().expect("endless WAL");
+                ProposerAcceptorMessage::AppendRequest(AppendRequest {
+                    h: AppendRequestHeader {
+                        term: 1,
+                        term_start_lsn: Lsn(0),
+                        begin_lsn: lsn,
+                        end_lsn: lsn + record.len() as u64,
+                        commit_lsn: Lsn(0),
+                        truncate_lsn: Lsn(0),
+                        proposer_uuid: [0; 16],
+                    },
+                    wal_data: record,
+                })
+            },
+            // Benchmark message processing (time per message).
+            |msg| {
+                runtime
+                    .block_on(safekeeper.process_msg(msg))
+                    .expect("message failed")
+            },
+            BatchSize::SmallInput, // automatically determine a batch size
+        );
+        Ok(())
+    }
+}
+
+/// Benchmarks WalAcceptor by sending it a batch of WAL records and waiting for it to confirm that
+/// the last LSN has been flushed to storage. We pipeline a bunch of messages instead of measuring
+/// each individual message to amortize costs (e.g. fsync), which is more realistic. Records are
+/// XlLogicalMessage with a tiny payload.
```

Comment:
A mention of how much data (in bytes) is ingested during this benchmark would be great.